### PR TITLE
feat: add virtualization object types support

### PR DESCRIPTION
**Pull request by leoparente**

### CODE REVIEW

File: diode-server/docker/sample.env [hunk 1]
```diff
@@ -5,7 +5,7 @@ REDIS_HOST=diode-redis
 REDIS_PORT=6378
 RECONCILER_GRPC_HOST=diode-reconciler
 RECONCILER_GRPC_PORT=8081
-NETBOX_DIODE_PLUGIN_API_BASE_URL=http://NETBOX_HOST/api/plugins/diode
+NETBOX_DIODE_PLUGIN_API_BASE_URL=http://192.168.31.114:8000/netbox/api/plugins/diode
```
Comment (mfiedorowicz) [inline diode-server/docker/sample.env]: Can we revert this change, please? It's a local variable IP

Comment (leoparente) [reply diode-server/docker/sample.env]:
```suggestion
NETBOX_DIODE_PLUGIN_API_BASE_URL=http://NETBOX_HOST/api/plugins/diode
```

File: diode-server/netbox/virtualization_wrappers.go [hunk 2]
```diff
@@ -0,0 +1,1680 @@
...
+			}
+		}
+
+	}
+
+	if reconciliationRequired {
+		vw.hasChanged = true
+		vw.objectsToReconcile = append(vw.objectsToReconcile, vw)
+	}
+
```
Comment (mfiedorowicz) [inline diode-server/netbox/virtualization_wrappers.go:650]: To ensure we don't have duplicated objects to reconcile, we should use `dedupObjectsToReconcile` func:
```suggestion

        dedupObjectsToReconcile, err := dedupObjectsToReconcile(vw.objectsToReconcile)
	if err != nil {
		return nil, err
	}
	vw.objectsToReconcile = dedupObjectsToReconcile

```

File: diode-server/netbox/virtualization_wrappers.go [hunk 3]
```diff
@@ -0,0 +1,1680 @@
...
+				vw.objectsToReconcile = append(vw.objectsToReconcile, &TagDataWrapper{Tag: t, hasParent: true})
+			}
+		}
+	}
+
+	if reconciliationRequired {
+		vw.hasChanged = true
+		vw.objectsToReconcile = append(vw.objectsToReconcile, vw)
+	}
+
```
Comment (mfiedorowicz) [inline diode-server/netbox/virtualization_wrappers.go:1193]: To ensure we don't have duplicated objects to reconcile, we should use `dedupObjectsToReconcile` func:
```suggestion

        dedupObjectsToReconcile, err := dedupObjectsToReconcile(vw.objectsToReconcile)
	if err != nil {
		return nil, err
	}
	vw.objectsToReconcile = dedupObjectsToReconcile

```

File: diode-server/netbox/virtualization_wrappers.go [hunk 4]
```diff
@@ -0,0 +1,1680 @@
...
```
Comment (mfiedorowicz) [inline diode-server/netbox/virtualization_wrappers.go:1443]: To ensure we don't have duplicated objects to reconcile, we should use `dedupObjectsToReconcile` func:
```suggestion

        dedupObjectsToReconcile, err := dedupObjectsToReconcile(vw.objectsToReconcile)
	if err != nil {
		return nil, err
	}
	vw.objectsToReconcile = dedupObjectsToReconcile

```

File: diode-server/netbox/virtualization_wrappers.go [hunk 5]
```diff
@@ -0,0 +1,1680 @@
...
```
Comment (mfiedorowicz) [inline diode-server/netbox/virtualization_wrappers.go:1675]: To ensure we don't have duplicated objects to reconcile, we should use `dedupObjectsToReconcile` func:
```suggestion

        dedupObjectsToReconcile, err := dedupObjectsToReconcile(vw.objectsToReconcile)
	if err != nil {
		return nil, err
	}
	vw.objectsToReconcile = dedupObjectsToReconcile

```

File: diode-server/reconciler/changeset/changeset_virt_test.go [hunk 6]
```diff
@@ -1138,12 +1138,7 @@ func TestVirtualizationPrepare(t *testing.T) {
 					queryParams:    map[string]string{"q": "undefined"},
 					objectChangeID: 0,
 					object: &netbox.DcimSiteDataWrapper{
-						Site: &netbox.DcimSite{
-							ID:     1,
-							Name:   "undefined",
-							Slug:   "undefined",
-							Status: (*netbox.DcimSiteStatus)(strPtr(string(netbox.DcimSiteStatusActive))),
-						},
```
Comment (mfiedorowicz) [inline diode-server/reconciler/changeset/changeset_virt_test.go]: Could we cover both use cases: with existing site and without, please?

Comment (mfiedorowicz) [reply diode-server/reconciler/changeset/changeset_virt_test.go]: Other than that, we should be good to go 🚀

Comment (leoparente) [reply diode-server/reconciler/changeset/changeset_virt_test.go]: done @mfiedorowicz  see last commit

**Review by mfiedorowicz — APPROVED** [review]
Thanks @leoparente, great work 🙇

File: diode-server/netbox/virtualization.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	VirtualizationClusterGroupObjectType = "virtualization.clustergroup"
+
+	// VirtualizationClusterTypeObjectType represents the Virtualization Cluster Type object type
+	VirtualizationClusterTypeObjectType = "virtualization.clustertype"
+
+	// VirtualizationVirtualMachineObjectType represents the Virtualization Virtual Machine object type
+	VirtualizationVirtualMachineObjectType = "virtualization.virtualmachine"
+
+	// VirtualizationInterfaceObjectType represents the Virtualization Interface object type
+	VirtualizationInterfaceObjectType = "virtualization.interface"
```
Comment (mfiedorowicz) [inline diode-server/netbox/virtualization.go]: Correct NetBox object type label for VM interface should be `virtualization.vminterface`, `virtualization.interface` results with:
```json
{"error": "ObjectType matching query does not exist.", "exception": "DoesNotExist", "netbox_version": "4.0.9", "python_version": "3.11.6"}
```

**Review by mfiedorowicz — COMMENTED** [review]
One more fix needed 🙏

File: diode-server/netbox/virtualization.go (hunk 7 above)
Comment (leoparente) [reply diode-server/netbox/virtualization.go]: Should I rename VirtualInterface to VMInterface in all places as well?

Comment (leoparente) [reply diode-server/netbox/virtualization.go]: to be more consistent

Comment (mfiedorowicz) [reply diode-server/netbox/virtualization.go]: Would be great, good suggestion!

Comment (leoparente) [reply diode-server/netbox/virtualization.go]: I've done the change and also updated SDK. Please, validate it again.

**Review by mfiedorowicz — APPROVED** [review]
LGTM 🚀

File: diode-server/reconciler/ingestion_processor.go [hunk 8]
```diff
@@ -346,6 +346,18 @@ func extractObjectType(in *diodepb.Entity) (string, error) {
 		return netbox.IpamIPAddressObjectType, nil
 	case *diodepb.Entity_Prefix:
 		return netbox.IpamPrefixObjectType, nil
+	case *diodepb.Entity_ClusterGroup:
+		return netbox.VirtualizationClusterGroupObjectType, nil
+	case *diodepb.Entity_ClusterType:
+		return netbox.VirtualizationClusterGroupObjectType, nil
```
Comment (mfiedorowicz) [inline diode-server/reconciler/ingestion_processor.go]: Should be `VirtualizationClusterTypeObjectType`

**Review by mfiedorowicz — CHANGES_REQUESTED** [review]
One fix needed to enable cluster type entity ingestion

File: diode-server/netbox/virtualization.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	Disk        *int                   `json:"disk,omitempty"`
+	Description *string                `json:"description,omitempty"`
+	Comments    *string                `json:"comments,omitempty"`
+	Tags        []*Tag                 `json:"tags,omitempty"`
+}
+
+// VirtualizationVMInterface represents a Virtualization Interface
+type VirtualizationVMInterface struct {
+	ID             int                           `json:"id,omitempty"`
+	VirtualMachine *VirtualizationVirtualMachine `json:"virtual_machine,omitempty"`
```
Comment (mfiedorowicz) [inline diode-server/netbox/virtualization.go]: Needs following change, otherwise virtual machine won't be unserialized properly
```suggestion
	VirtualMachine *VirtualizationVirtualMachine `json:"virtual_machine,omitempty"  mapstructure:"virtual_machine"`
```

File: diode-server/netbox/virtualization.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+	MTU            *int                          `json:"mtu,omitempty"`
+	MACAddress     *string                       `json:"mac_address,omitempty" mapstructure:"mac_address,omitempty"`
+	Description    *string                       `json:"description,omitempty"`
+	Tags           []*Tag                        `json:"tags,omitempty"`
+}
+
+// VirtualizationVirtualDisk represents a Virtualization Virtual Disk
+type VirtualizationVirtualDisk struct {
+	ID             int                           `json:"id,omitempty"`
+	VirtualMachine *VirtualizationVirtualMachine `json:"virtual_machine,omitempty"`
```
Comment (mfiedorowicz) [inline diode-server/netbox/virtualization.go]: Needs following change, otherwise virtual machine won't be unserialized properly
```suggestion
	VirtualMachine *VirtualizationVirtualMachine `json:"virtual_machine,omitempty"  mapstructure:"virtual_machine"`
```

**Review by mfiedorowicz — COMMENTED** [review]
A couple of issues discovered while testing Go SDK with virtualization types support.

File: diode-server/reconciler/ingestion_processor.go (hunk 8 above)
Comment (leoparente) [reply diode-server/reconciler/ingestion_processor.go]:
```suggestion
		return netbox.VirtualizationClusterTypeObjectType, nil
```

**Review by mfiedorowicz — APPROVED** [review]

**Review by mfiedorowicz — APPROVED** [review]
